frontend/lib/resolution/default-functions.cpp

```diff
@@ -488,6 +493,116 @@ const TypedFnSignature* fieldAccessor(Context* context,
   return fieldAccessorQuery(context, compType, fieldName);
 }
 
+// generate formal detail and formal type
+static void
+generateOperatorFormalDetail(const UniqueString name,
+                             const CompositeType*& compType,
+                             std::vector<UntypedFnSignature::FormalDetail>& ufsFormals,
+                             std::vector<QualifiedType>& formalTypes,
+                             QualifiedType::Kind qtKind,
+                             bool hasDefault = false,
+                             const uast::Decl* decl = nullptr) {
+  auto fd = UntypedFnSignature::FormalDetail(name, hasDefault, decl);
+  ufsFormals.push_back(std::move(fd));
+
+  auto qtFd = QualifiedType(qtKind, compType);
+  formalTypes.push_back(std::move(qtFd));
+}
+
+// builds the formal entries for the operator methods, including the 'this'
+// method receiver and the lhs and rhs arguments. Specify the
+// QualifiedType::Kind for each of these.
+static void
+generateOperatorMethodParts(Context* context,
+                            const CompositeType* inCompType,
+                            const CompositeType*& compType,
+                            std::vector<UntypedFnSignature::FormalDetail>& ufsFormals,
+                            std::vector<QualifiedType>& formalTypes,
+                            QualifiedType::Kind thisKind,
+                            QualifiedType::Kind lhsKind,
+                            QualifiedType::Kind rhsKind) {
+  // adjust to refer to fully generic signature if needed
+  auto genericCompType = inCompType->instantiatedFromCompositeType();
+  compType = genericCompType ? genericCompType : inCompType;
+
+  // make sure the receiver is a record type
+  CHPL_ASSERT(compType->isRecordType() && "Only RecordType supported for now");
+
+  // start by adding a formal for the receiver, 'this'
+  generateOperatorFormalDetail(USTR("this"), compType, ufsFormals, formalTypes,
+                               thisKind);
+
+  // add a formal for the 'lhs' argument
+  generateOperatorFormalDetail(UniqueString::get(context,"lhs"),
+                               compType, ufsFormals, formalTypes,
+                               lhsKind);
+
+  // add a formal for the 'rhs' argument
+  generateOperatorFormalDetail(UniqueString::get(context,"rhs"),
+                               compType, ufsFormals, formalTypes,
+                               rhsKind);
+
+  CHPL_ASSERT(formalTypes.size() == 3);
+  CHPL_ASSERT(ufsFormals.size() == 3);
+}
+
+/*
+generate a TypedFnSignature and UntypedFnSignature with formal details for a
+record operator method. The operator is specified by the UniqueString op.
+*/
+static const TypedFnSignature*
+generateRecordOperator(Context* context, UniqueString op,
```

Comment:
IMO it would be better structure to have a different function for each of the things we are generating; e.g. `generateRecordAssignment` and `generateRecordEquality`. The structure you have here won't be smooth when we add things like `hash` which don't have both an `lhs` and an `rhs`. I'd recommend creating those different routines and factoring out common code into helper functions that each of these routines can call. (see my other comment below)